internal/runner/runnerpool/builder_helpers.go

```diff
@@ -0,0 +1,238 @@
+}
+
+// newBaseDiscovery constructs the base discovery with common immutable options.
+func newBaseDiscovery(
+	tgOpts *options.TerragruntOptions,
+	workingDir string,
+	configFilenames []string,
+	opts ...common.Option,
+) *discovery.Discovery {
+	anyOpts := make([]interface{}, len(opts))
+	for i, v := range opts {
+		anyOpts[i] = v
+	}
+
+	return discovery.
+		NewDiscovery(workingDir).
+		WithOptions(anyOpts...).
+		WithDiscoverExternalDependencies().
+		WithParseInclude().
+		WithParseExclude().
+		WithDiscoverDependencies().
+		WithSuppressParseErrors().
+		WithConfigFilenames(configFilenames).
+		WithDiscoveryContext(&component.DiscoveryContext{
+			Cmd:  tgOpts.TerraformCliArgs.First(),
+			Args: tgOpts.TerraformCliArgs.Tail(),
+		})
+}
+
+// prepareDiscovery constructs a configured discovery instance based on Terragrunt options and flags.
+func prepareDiscovery(
+	tgOpts *options.TerragruntOptions,
+	excludeByDefault bool,
+	opts ...common.Option,
+) (*discovery.Discovery, error) {
+	workingDir := resolveWorkingDir(tgOpts)
+	configFilenames := buildConfigFilenames(tgOpts)
+
+	d := newBaseDiscovery(tgOpts, workingDir, configFilenames, opts...)
+
+	// Include / exclude directories
+	if len(tgOpts.IncludeDirs) > 0 {
+		d = d.WithIncludeDirs(tgOpts.IncludeDirs)
+	}
+
+	if len(tgOpts.ExcludeDirs) > 0 {
+		d = d.WithExcludeDirs(tgOpts.ExcludeDirs)
+	}
+
+	// Include behavior flags
+	if tgOpts.StrictInclude {
+		d = d.WithStrictInclude()
+	}
+
+	if excludeByDefault {
+		d = d.WithExcludeByDefault()
+	}
+
+	// Enable reading file tracking when requested by CLI flags
+	if len(tgOpts.ModulesThatInclude) > 0 || len(tgOpts.UnitsReading) > 0 {
+		d = d.WithReadFiles()
+	}
+
+	// Apply filter queries when provided
+	if len(tgOpts.FilterQueries) > 0 {
+		filters, err := parseFilters(tgOpts.FilterQueries, workingDir)
+		if err != nil {
+			return nil, errors.Errorf("failed to parse filter queries in %s: %w", workingDir, err)
+		}
+
+		d = d.WithFilters(filters)
+	}
+
+	return d, nil
+}
+
+// runDiscovery executes discovery with telemetry and returns discovered components.
+func runDiscovery(
+	ctx context.Context,
+	l log.Logger,
+	d *discovery.Discovery,
+	tgOpts *options.TerragruntOptions,
+) (component.Components, error) {
+	var discovered component.Components
+
+	err := doWithTelemetry(ctx, telemetryDiscovery, map[string]any{
+		"working_dir":       tgOpts.WorkingDir,
+		"terraform_command": tgOpts.TerraformCommand,
+	}, func(childCtx context.Context) error {
+		var discoveryErr error
+
+		discovered, discoveryErr = d.Discover(childCtx, l, tgOpts)
+		if discoveryErr == nil {
+			l.Debugf("Runner pool discovery found %d configs", len(discovered))
+		}
+
+		return discoveryErr
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	return discovered, nil
+}
+
+// maybeRetryDiscovery retries discovery without exclude-by-default if none were discovered
+// and modules-that-include / units-reading flags are set.
+func maybeRetryDiscovery(
```

Comment:
is there a way this could be eliminated, and instead the caller of this could do the check, remove the option, and re-run `runDiscovery`?

Alternatively, could this be refactored to do the check, remove the option, and run `runDiscovery` internally rather than duplicating the logic?